server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -52,25 +52,295 @@
 
 public class VariableWidthHistogramAggregator extends DeferableBucketAggregator {
 
+    private abstract class CollectionMode{
+        abstract void collectValue(LeafBucketCollector sub, int doc, double val) throws IOException;
+    }
+
+    private class CacheValues extends CollectionMode{
+        public int numCachedDocs;
+        public DoubleArray cachedValues;
+
+        CacheValues(){
+            this.numCachedDocs = 0;
+            this.cachedValues = bigArrays.newDoubleArray(1);
+        }
+
+        @Override
+        public void collectValue(LeafBucketCollector sub, int doc, double val) throws IOException{
+            // Cache the doc in a new bucket, rather than trying to add it into an existing bucket
+            cachedValues = bigArrays.grow(cachedValues, numCachedDocs + 1);
+            cachedValues.set((long) numCachedDocs, val);
+            collectBucket(sub, doc, numCachedDocs);
+            numCachedDocs += 1;
+        }
+
+        public void releaseCache(){
```

Comment:
Better to have the superclass `implement Releasable`, I think. Those jump out at me real quick and I'm used to keeping a mental list of what has to be released.